Add KubeCon2019/Tom Keynote blog post (#168)
grafana/grafana.org: content/blog/2019-05-27-grafana-labs-at-kubecon-what-is-the-future-of-observability.md

"The three pillars were really meant as a framework for people who have just gotten started on their journey in observability. In 2018 there's been a conversation -- and some critique -- about this," said Branczyk.

"We have all of this data, and we're telling people if you have metrics, if you have logs, and if you have tracing you've solved observability. You have an observable system!" Branczyk explained. "But we don't think that's the case. There is so much more for observability to come."

"It’s always a bit of a risky business doing predictions," said Wilkie. "But we’re going to give it a go anyway." 

Below is a recap of their [Kubecon keynote](https://youtu.be/MkSdvPdS1oA).

## The Three Pillars
To start, here’s a quick overview of each pillar.

![Pillars Slide](/assets/img/blog/kubecon2019_tomkeynote_pillars.png)

**1. Metrics**
Normally this is time series data that is used for trends in memory usage and latency. 

"The CNCF has some great projects in this space," said Wilkie. "OpenMetrics is an exposition format for exporting metrics from your application, and Prometheus is probably now the defacto monitoring system for Kubernetes and apps on Kubernetes."

**2. Logs**

> Logs, or events, are what comes out of your containers in Kubernetes. The CNCF has Fluentd and there’s log shipping.

Reviewer: ```suggestion
Logs, or events, are what comes out of your containers on "standard out" in Kubernetes. Think error messages, exceptions and request logs. The CNCF has the Fluentd project, which is a log shipping agent.
```

In this space, there is OpenTelemetry as well as Jaeger, a CNCF project which Grafana Labs utilizes, according to Wilkie. 

## Prediction #1: More Correlation Between Pillars
"The first prediction is that there will be more correlation between the different pillars," said Wilkie. "We think this is the year when we're going to start breaking down the walls, and we're going to start seeing joined up workflows."

Here are three examples of workflows and projects that provide automated correlation that you can do today: 

![Correlation Slide](/assets/img/blog/kubecon2019_tomkeynote_correlation.png)

**1. Grafana Loki**
"The first system is actually a project that I work on myself called [Loki](https://grafana.com/blog/2018/12/12/loki-prometheus-inspired-open-source-logging-for-cloud-natives/)," said Wilkie, who also delivered [a separate KubeCon talk](https://grafana.com/blog/2019/05/23/grafana-labs-at-kubecon-lokis-march-toward-ga/) about the open source log aggregation system that Grafana Labs launched six months ago. Since then "we have had an absolutely great response. Loads of people have given us really good feedback." 

> Loki uses Prometheus' service discovery to automatically find jobs within your cluster. It then takes that service discovery as well as the labels that the service discovery gives you and associates them with the log stream, preserving its context and [saving you money](https://grafana.com/blog/2019/05/06/how-loki-correlates-metrics-and-logs----and-saves-you-money/). 

Reviewer: ```suggestion
Loki uses Prometheus' service discovery to automatically find jobs within your cluster. It then takes the labels that the service discovery gives you and associates them with the log stream, preserving context and [saving you money](https://grafana.com/blog/2019/05/06/how-loki-correlates-metrics-and-logs----and-saves-you-money/). 
```

"It's this kind of systematic, consistent metadata that's the same between your logs and your metrics that enables switching between the two seamlessly," said Wilkie. 

**2. Elasticsearch & Zipkin**
"Elasticsearch is probably the most popular log aggregation system, even I can admit that," said Wilkie. "And Zipkin is the original open source distributed tracing system." 

Within Kibana, the Elasticsearch UI, there is a function called field formatters which allows someone to hyperlink between tracing and logs. 

"Some chap on Twitter set up his Kibana to insert a link using a field formatter so that he could instantly link to his Zipkin traces," said Wilkie. "I think this is really cool, and I'm really looking forward to adding this kind of feature to Grafana." 

**3. OpenTelemetry**
Recently Google’s OpenCensus and CNCF’s OpenTracing merged into one open source project known as [OpenTelemetry](https://thenewstack.io/opentracing-opencensus-merge-into-a-single-new-project-opentelemetry/) now operated by CNCF. 

> Within OpenTracing, the use of exemplars illustrates another example of correlation. "[Exemplars] are example trace IDs associated with every bucket in a histogram," said Wilkie. "So you can see what’s causing high latency and link straight to the trace." 

Reviewer: ```suggestion
Within OpenCensus, the use of exemplars illustrates another example of correlation. "Exemplars are trace IDs associated with every bucket in a histogram," said Wilkie. "So you can see what’s causing high latency and link straight to the trace." 
```

"I like that OpenTelemetry is open source," added Wilkie. "I'm not actually aware of an open source system on the server side that has implemented this workflow. If I'm wrong, come and find me." (He’s [@tom_wilkie](https://twitter.com/tom_wilkie) on Twitter!)

## Prediction #2: New Signals & New Analysis

> Your o11y toolkit doesn’t have to include just three pillars. 

Reviewer: ```suggestion
Your observability toolkit doesn’t have to include just three pillars. 
```

(Don't use o11y please)

"As humans we like the number three. That's why maybe we've settled on the three pillars of observability, but we believe there is so much more data that can help us have insight into running systems," said Branczyk.

The question is: How do we make all this data useful? The answer is in the metadata. 

> "Metadata is what's going to allow us to do all of this correlation but still build systems that are useful individually. Also as a whole, it will do all this great correlation across the signals," said Branczyk. 

Reviewer: I'd drop these last two paragraphs, they seem out of place.

"I think we're going to be seeing a lot more signals and analysis," concluded Branczyk. "I've only shown you one example but I think there's going to be a lot more out there to explore."

## Prediction #3: Rise of Index-Free Log Aggregation
Over the past six months to a year, the feelings around log aggregation have been mutual. "A lot of people have been saying things like just give me log files and grep," said Wilkie.  

"The systems like Splunk and Elasticsearch give us a tremendous amount of power to search and analyze our logs but all of this power comes with a lot of ... I'm not going to say responsibility. It comes with a lot of complexity -- and expense," said Wilkie. 

Before Splunk and Elasticsearch, logs were just stored as files on disk, possibly in a centralized server, and it was easy to go on and grep it. “I think we're starting to see the desire for simpler index-free log aggregation systems,” said Wilkie. “Effectively everything old is new again.” 

Here Wilkie gives three examples ("Again, three is a very aesthetically pleasing number," he joked) of how this works:  

![Logs Slide](/assets/img/blog/kubecon2019_tomkeynote_logs.png)

**1. OK Log**

> Peter Balkan started this project over a year ago but "unfortunately it's been discontinued," said Wilkie. "But it had some really great ideas about distributing grep over a cluster of machines and being able to basically just brute force your way through your logs. It made it a lot easier to operate and a lot cheaper to run."

Reviewer: Peter Bourgon

Link to https://github.com/oklog/oklog
